### ISSUE TITLE
Container: "Unable to watch for file changes in this large workspace"

### ISSUE THREAD

**Issue by dfarley1** [microsoft/vscode-remote-release#3804]
- VSCode Version: 1.49.3
- Local OS Version: Ubuntu 20.04
- Remote OS Version: Gentoo
- Remote Extension/Connection Type: Container on Remote Host

Steps to Reproduce:

1. Connect to the docker container.
2. Open our repo workspace file. Single-root workspace at `/depot` in the container.
3. See the notification linking to https://code.visualstudio.com/docs/setup/linux#_visual-studio-code-is-unable-to-watch-for-file-changes-in-this-large-workspace-error-enospc

Our container seems to have enough watchers for the number of files in our folder, so I'm confused why we're seeing this error:
```
dan@docker /depot $ cat /proc/sys/fs/inotify/max_user_watches
524288
dan@docker /depot $ sudo find /depot -type f,d,l,p,s -print | wc -l
53358
```

**Comment by chrmarti** [comment]
What are the `max_user_watches` on your Linux running the Docker daemon? Try increasing these, does that fix it?

**Comment by PavelSosin-320** [comment]
Indeed, Kernel parameters [Kernel parameters: user watches: ](https://gist.github.com/tsrivishnu/9f551ef0098021a913e01d6d594c555d) Container has no own kernel which can be configured during build! It uses the kernel of the Host on which Docker runs. Always check your Host resources 1st
To run big container the very big Host is required! The Docker containers share host resources according to cgroups
And allocate enough resources (cpu, etc.) **when you run the container**.

**Comment by dfarley1** [comment]
> What are the `max_user_watches` on your Linux running the Docker daemon? Try increasing these, does that fix it?

I get the same max value on the Docker's host: `524288`.

> To run big container the very big Host is required! The Docker containers share host resources according to cgroups

There is only 1 container running on the host and it is not doing much else aside from hosting the container.

**Comment by PavelSosin-320** [comment]
@chrmarti Sorry for intervention into your internal discussion but it is connected to another issue - sharing git configuration among dev container and local Windows and WSL environment: if crlf = true and configuration is not shared files committed to git from Windows or WSL can't be checked from git as is in WSL, for example. Will it be an issue?  
The average cost for me was 1 week per feature involving tokenizer on debugging and test development.

**Comment by dfarley1** [comment]
I think I found at least part of the reason. It appears that VSCode is leaving a bunch of child processes running from previous instances. This is a "long running" container, I think it was up for a couple months.

Without any instances of VSCode attached to the container there were over 200 instances of `/home/dan/.vscode-server/bin/<hash>/node` running in the background; 21 of them appear to be related to file watching:
```
dan 12633  0.0  0.1 918216 51248 ?        Sl   Sep01  51:05 /home/dan/.vscode-server/bin/<hash>/node /home/dan/.vscode-server/bin/<hash>/out/bootstrap-fork --type=watcherService
```
Most of the others are running JS scripts/snippets. I can provide the full log privately if you wish.

I suspect that the VSCode server isn't properly cleaning itself up in some cases. I'm working over a VPN to my workstation in an office, so things might not get cleaned up correctly if there's a "harsh" disconnection like if I flip the VPN off or shut my laptop without closing VSCode first.

**Comment by chrmarti** [comment]
Fixed an issue that would explain this with #3850. Please check with the next version (0.150.0) if this is fixed.

**Comment by rzhao271** [comment]
@dfarley1 wondering if you can see whether the issue is fixed for you in the latest Insiders? https://code.visualstudio.com/insiders/

It has a version of the extension that contains the fix
